tokio/src/io/util/read_to_end.rs

```diff
@@ -67,20 +67,41 @@ fn poll_read_to_end<V: VecU8, R: AsyncRead + ?Sized>(
     // has 4 bytes while still making large reads if the reader does have a ton
     // of data to return. Simply tacking on an extra DEFAULT_BUF_SIZE space every
     // time is 4,500 times (!) slower than this if the reader has a very small
-    // amount of data to return.
-    buf.reserve(32);
+    // amount of data to return. When the vector is full with its starting
+    // capacity, we first try to read into a small buffer to see if we reached
+    // an EOF. This avoids the unnecessary allocation that we attempt before
+    // reading into the vector.
+
+    const NUM_BYTES: usize = 32;
+    let try_small_read = buf.try_small_read_first(NUM_BYTES);
 
     // Get a ReadBuf into the vector.
     let mut read_buf = buf.get_read_buf();
-
     let filled_before = read_buf.filled().len();
-    let poll_result = read.poll_read(cx, &mut read_buf);
-    let filled_after = read_buf.filled().len();
-    let n = filled_after - filled_before;
+    let poll_result;
+
+    let filled_after = if try_small_read {
+        let mut small_buf: [MaybeUninit<u8>; NUM_BYTES] = [MaybeUninit::uninit(); NUM_BYTES];
+        let mut small_read_buf = ReadBuf::uninit(&mut small_buf);
+        poll_result = read.poll_read(cx, &mut small_read_buf);
+        let filled = small_read_buf.filled();
+        if filled.len() > read_buf.remaining() {
+            buf.reserve(NUM_BYTES);
+            read_buf = buf.get_read_buf();
+            read_buf.put_slice(filled);
+        }
```

Comment:
Now you're throwing away the bytes completely if `filled.len() <= read_buf.remaining()`.

How about a test along these lines?
```Rust
#[tokio::test]
async fn all_combinations() {
    let arr: Vec<u8> = (0..100).collect();

    for len in 1..100 {
        let bytes = &arr[..i];
        for split in 0..len {
            for cap in 0..101 {
                let mut mock = if split == 0 {
                    Builder::new().read(bytes)
                } else {
                    Builder::new().read(&bytes[..split]).read(&bytes[split..])
                };
                let mut buf = Vec::with_capacity(cap);
                AsyncReadExt::read_to_end(&mut mock, &mut buf)
                    .await
                    .unwrap();
                // It has the right data.
                assert_eq!(buf.as_bytes(), bytes);
                // Unless cap was smaller than length, then we did not reallocate.
                if cap >= len {
                    assert_eq!(buf.capacity(), cap);
                }
            }
        }
    }
}
```

Reply:
Oops, I forgot to push my change that moved put_slice outside the if statement. I did add your test that caught that. I start from len from 32 however, since the code allocates at most 32 bytes each time.